### PR TITLE
Aaf

### PR DESCRIPTION
**Pull request by cbedgren**
Created dataset and data loader for AAF dataset

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -6,6 +6,9 @@ logs
 .vscode
 chkp/
 ignore/
+.venv/
+**/*.jpg
```
Comment (spencermyoung513) [inline .gitignore]: I'm wary of excluding all images from being tracked in the repo. Let's just make sure images get put in an ignored directory if we don't want to commit them

File: configs/aaf_config.yaml [hunk 2]
```diff
@@ -0,0 +1,28 @@
+experiment_name: aaf
+random_seed: 1998
+head_type: gaussian
+dataset:
+  type: image
+  spec: aaf
+  path: aaf
```
Comment (spencermyoung513) [inline configs/aaf_config.yaml]: You can leave the `path` key empty since the `spec` is indicated

File: probcal/custom_datasets/aaf_dataset.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+from PIL import Image
+
+
+from torch.utils.data import Dataset
+
+class AAFDataset(Dataset):
+    """The All-Age-Faces (AAF) Dataset contains 13'322 face images (mostly Asian) 
+    distributed across all ages (from 2 to 80), including 7381 females and 5941 males."""
+
+    def __init__(self, csv_file, root_dir, transform=None):
```
Comment (spencermyoung513) [inline probcal/custom_datasets/aaf_dataset.py]: Go ahead and toss type annotations onto these (e.g. `csv_file: str`, `root_dir: str`, `transform: typing.Callable`). Should help catch more bugs

File: probcal/data_modules/aaf_datamodule.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+
+class AAFDataModule(L.LightningDataModule):
+
+    IMG_SIZE = 224
+
+    def __init__(
+        self,
+        csv_file_path: str | Path,
+        root_dir: str | Path,
```
Comment (spencermyoung513) [inline probcal/data_modules/aaf_datamodule.py]: It would be nice if this class took care of the downloading / CSV generation automatically. So in theory, all a new user would have to do is instantiate the AAFDataModule, call `prepare_data`, and everything would be set up. The COCO People dataset /datamodule is a good example of how this could look

File: probcal/data_modules/aaf_datamodule.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        super().__init__()
+        self.csv_file_path = csv_file_path
+        self.root_dir = Path(root_dir)
+        self.batch_size = batch_size
+        self.num_workers = num_workers
+        self.persistent_workers = persistent_workers
+        self.surface_image_path = surface_image_path
+
+    def prepare_data(self) -> None:
+        # Force images to be downloaded.
```
Comment (spencermyoung513) [inline probcal/data_modules/aaf_datamodule.py]: As far as I can tell, creating the `AAFDataset` doesn't force images to be downloaded, right?

File: probcal/enums.py [hunk 6]
```diff
@@ -11,7 +11,8 @@ class AcceleratorType(Enum):
 class HeadType(Enum):
     GAUSSIAN = "gaussian"
     POISSON = "poisson"
-    NEGATIVE_BINOMIAL = "negative_binomial"
+    NEGATIVE_BINOMIAL = "nbinom"
```
Comment (spencermyoung513) [inline probcal/enums.py]: Any specific reason we changed the value for `NEGATIVE_BINOMIAL`? This could break our support for older configs

File: probcal/enums.py [hunk 7]
```diff
@@ -32,6 +33,9 @@ class DatasetType(Enum):
 class ImageDatasetName(Enum):
     MNIST = "mnist"
     COCO_PEOPLE = "coco_people"
+    CIFAR10 = 'cifar_10'
+    CIFAR100 = 'cifar_100'
```
Comment (spencermyoung513) [inline probcal/enums.py]: Let's leave this out unless we have those datasets implemented

File: probcal/utils/experiment_utils.py [hunk 8]
```diff
@@ -74,6 +78,14 @@ def get_datamodule(
                 num_workers=num_workers,
                 persistent_workers=True if num_workers > 0 else False,
             )
+        elif dataset_spec == ImageDatasetName.AAF:
+            return AAFDataModule(
+                csv_file_path = os.path.join(GLOBAL_DATA_DIR, "aaf/image_sets/picture_data.csv"),
+                root_dir=os.path.join(GLOBAL_DATA_DIR, "aaf/original_images"),
```
Comment (spencermyoung513) [inline probcal/utils/experiment_utils.py]: As things currently stand, the CSV file path doesn't exist in other people's copies of the repo. And `root_dir` also doesn't exist by default

File: probcal/utils/configs.py [hunk 9]
```diff
@@ -117,6 +117,8 @@ def from_yaml(config_path: str | Path) -> TrainingConfig:
         dataset_type = DatasetType(config_dict["dataset"]["type"])
         if dataset_type == DatasetType.TABULAR:
             dataset_path = Path(config_dict["dataset"]["path"])
+        elif dataset_type == DatasetType.IMAGE:
+            dataset_path = ImageDatasetName(config_dict["dataset"]["path"])
```
Comment (spencermyoung513) [inline probcal/utils/configs.py]: Porter is making some upstream changes to sort out a better way to handle `dataset_path` when there isn't a specific file that defines the dataset like in the tabular case. Let's wait on that and then I can provide some guidance for how to handle this

**Review by spencermyoung513 — COMMENTED** [review]
Looking good! I left a few comments that will hopefully help. I'm gonna work on a few structural things for the repo that should make this a bit easier, and also add some pre-commit linters and whatnot to unify how all the code gets formatted

File: .gitignore (hunk 1 above)
Comment (cbedgren) [reply .gitignore]: Sounds good!

File: probcal/enums.py (hunk 6 above)
Comment (cbedgren) [reply probcal/enums.py]: On some of the code that I was looking at I saw "nbinom" used, I can make sure that it gets switched back to "negative_binomial"

Comment (spencermyoung513) [reply probcal/enums.py]: Yeah, I think nbinom is used in some filenames, but the value for configs should be `negative_binomial`. Probably could be cleaned up in the future haha

File: probcal/data_modules/aaf_datamodule.py (hunk 4 above)
Comment (cbedgren) [reply probcal/data_modules/aaf_datamodule.py]: Okay! We will work on implementing this

File: .gitignore (hunk 1 above)
Comment (andrewhall1124) [reply .gitignore]: Fixed: removed venv and jpg from gitignore.

File: configs/aaf_config.yaml (hunk 2 above)
Comment (andrewhall1124) [reply configs/aaf_config.yaml]: Fixed.

File: probcal/custom_datasets/aaf_dataset.py (hunk 3 above)
Comment (andrewhall1124) [reply probcal/custom_datasets/aaf_dataset.py]: Fixed

File: probcal/data_modules/aaf_datamodule.py (hunk 4 above)
Comment (andrewhall1124) [reply probcal/data_modules/aaf_datamodule.py]: Fixed

File: probcal/data_modules/aaf_datamodule.py (hunk 5 above)
Comment (andrewhall1124) [reply probcal/data_modules/aaf_datamodule.py]: Fixed

File: probcal/enums.py (hunk 7 above)
Comment (andrewhall1124) [reply probcal/enums.py]: Fixed.

File: probcal/utils/experiment_utils.py (hunk 8 above)
Comment (andrewhall1124) [reply probcal/utils/experiment_utils.py]: Fixed

File: probcal/utils/configs.py (hunk 9 above)
Comment (andrewhall1124) [reply probcal/utils/configs.py]: Okie dokie

File: configs/aaf_config.yaml [hunk 10]
```diff
@@ -0,0 +1,27 @@
+experiment_name: aaf
+random_seed: 1998
+head_type: gaussian
+dataset:
+  type: image
+  spec: aaf
+  input_dim: 3
```
Comment (spencermyoung513) [inline configs/aaf_config.yaml]: This key is only used when the data being passed in is tabular (since the MLP we build has to have the same number of nodes in the first layer as features in the data). So you can drop `input_dim` for image data

Comment (andrewhall1124) [reply configs/aaf_config.yaml]: Fixed

**Comment by spencermyoung513** [comment]
To keep the repo / commit history clean, could you do the following two things before merging?

1. Squash all your commits into one, then rebase onto `main` so we only add one commit to the repo history. You can accomplish this with `git reset <commit-sha>` (choose the commit sha based on the last commit from main that's in your git history before your branching changes occurred) and then re-staging all your changes and committing them once.
2. Run the linters on your changes (see the pre-commit section in README.md in the latest version of `main`) to auto-format and standardize imports, catch small bugs, etc.

File: .gitignore [hunk 11]
```diff
@@ -6,6 +6,7 @@ logs
 .vscode
 chkp/
 ignore/
+data
```
Comment (spencermyoung513) [inline .gitignore]: Is this necessary if we already have `data/`?

File: configs/train/aaf/double_poisson_aaf_cfg.yaml [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+random_seed: 1998
+head_type: double_poisson
+dataset:
+  type: image
+  spec: aaf
+training:
+  accelerator: gpu
+  batch_size: 32
+  chkp_dir: chkp
+  chkp_freq: 6
```
Comment (spencermyoung513) [inline configs/train/aaf/double_poisson_aaf_cfg.yaml]: Does this mean you saved a model every 6 epochs for 200 epochs? We can probably delete a lot of those intermediate ones haha

File: configs/train/aaf/double_poisson_aaf_cfg.yaml [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+  precision: bf16-mixed
+  optimizer:
+    type: adam_w
+    kwargs:
+      lr: 0.001
+      weight_decay: 0.00001
+  lr_scheduler:
+    type: cosine_annealing
+    kwargs:
+      T_max: 30
```
Comment (spencermyoung513) [inline configs/train/aaf/double_poisson_aaf_cfg.yaml]: This should match `num_epochs`. Otherwise you anneal to a basically-zero learning rate by the end of the 30th epoch, which could affect training dramatically

File: probcal/enums.py [hunk 14]
```diff
@@ -13,6 +13,7 @@ class HeadType(Enum):
     POISSON = "poisson"
     DOUBLE_POISSON = "double_poisson"
     NEGATIVE_BINOMIAL = "negative_binomial"
+    MULTI_CLASS = "multi_class"
```
Comment (spencermyoung513) [inline probcal/enums.py]: We can probably drop this `HeadType` since we're not training any `MultiClassNN` models

File: probcal/utils/configs.py [hunk 15]
```diff
@@ -6,7 +6,7 @@
 
 from probcal.enums import AcceleratorType
 from probcal.enums import BetaSchedulerType
-from probcal.enums import DatasetType
+from probcal.enums import DatasetType, ImageDatasetName
```
Comment (spencermyoung513) [inline probcal/utils/configs.py]: Double import based on line 11

File: configs/train/aaf/double_poisson_aaf_cfg.yaml (hunk 12 above)
Comment (spencermyoung513) [reply configs/train/aaf/double_poisson_aaf_cfg.yaml]: Looking at `ava`'s storage, it's at 98% capacity now. My guess is you have a ton of auxiliary models taking up storage space now. Delete everything you don't need, then run `df -h` and see if you can get the number down for `/dev/nvme0n1p2`

Comment (andrewhall1124) [reply configs/train/aaf/double_poisson_aaf_cfg.yaml]: What would you like the frequency to be? I think we just copied the coco config that you sent us.

Comment (spencermyoung513) [reply configs/train/aaf/double_poisson_aaf_cfg.yaml]: Have it match the number of epochs. The best model w.r.t. loss already saves separately, so we just need the one from the last epoch from the frequency argument

**Review by spencermyoung513 — APPROVED** [review]
